parcels/collection/collectionaos.py

```diff
@@ -484,6 +533,206 @@ def add_same(self, same_class):
                 p._cptr = pdata
         self._ncount = self._data.shape[0]
 
+    def add_multiple(self, data_array):
+        """
+        Add multiple particles from an array-like structure (i.e. list or tuple or np.ndarray)
+        to the collection.
+        :arg data_array: one of the following:
+            i) a list or tuples containing multple Particle instances
+            ii) a Numpy.ndarray of dtype = Particle dtype
+            iii) a dict of Numpy.ndarray of shape, each of which with N = # particles
+        :returns vector-list of indices of all added particles
+        """
+        super().add_multiple(data_array)
+        results = []
+        if len(data_array) <= 0:
+            return results
+        if isinstance(data_array, list) or isinstance(data_array, tuple):
+            logger.info("SoA -> data_array - type: {}; values: {}".format(type(data_array[0]), data_array))
+            for item in data_array:
+                results.append(self.add_single(item))
+        elif isinstance(data_array, np.ndarray) and (data_array.dtype == self._ptype):
+            for i in range(data_array.shape[0]):
+                pdata = data_array[i]
+                results.append(self.add_single(pdata))
+        elif isinstance(data_array, dict) and isinstance(data_array['lon'], np.ndarray):
+            pu_ids = None
+            pu_indices = None
+            n_pu_data = 0
+            if MPI and MPI.COMM_WORLD.Get_size() > 1:
+                mpi_comm = MPI.COMM_WORLD
+                mpi_size = mpi_comm.Get_size()
+                mpi_rank = mpi_comm.Get_rank()
+                spdata = np.array([data_array['lon'], data_array['lat']]).transpose([1, 0])
+                min_pu = None
+                if mpi_rank == 0:
+                    dists = distance.cdist(spdata, self._pu_centers)
```

Comment:
the `scipy.spatial.distance` methods works with respect to any of the meshes. The caveat, indeed, is that for `mesh=sperical` and a periodic boundary condition, the _wrapping_ of the coordinates is not accounted for. That said, a coordinate-wrapping in line with that scenario is also not supported by any data-driven segmentation method that is _off-the-shelf_ available. In short: it works for the purpose it is designed for: keep track of cluster centers established from `KMeans` (and alike segmentation algorithms) to facilitate quick, in-runtime association of new particles to one of the MPI processes, without rebuilding the tree on every `insert`.

Improvements about that are the key part of load-balancing in MPI environments, which is one of the key future features in the pipeline. The distance computation can indeed be made more coherent, so that one and the same mechanism can be used for MPI process association, ID generation and tree-parsing of particles for interaction kernels. This will be, though, a longer discussion on how to do it and goes beyond the scope of this PR. Again, it will be an inherent sub-challenge in the MPI load balancing, so we will come back to the point in either case in the future.

Reply:
OK, I created #1117 to follow up on this discussion